applications/FluidDynamicsApplication/custom_utilities/element_size_calculator.cpp

```diff
@@ -178,6 +178,241 @@ double ElementSizeCalculator<3,8>::MinimumElementSize(const Geometry<Node<3> >&
+
+    double Hvel = 0.0;
+
+    const unsigned int NumNodes = 4;
+
+    // Loop over edges looking for maximum 'projected' length
+    array_1d<double,3> Edge(3,0.0);
+    for(unsigned int i = 0; i < NumNodes; ++i)
+    {
+        for(unsigned int j = i+1; j < NumNodes; ++j)
+        {
+            double lu;
+            Edge = rGeometry[j] - rGeometry[i];
+            lu = rVelocity[0] * Edge[0];
+            for (unsigned int d = 1; d < 3; ++d)
+                lu += rVelocity[d] * Edge[d];
+            lu = fabs(lu);
+            if(Hvel < lu) Hvel = lu;
+        }
+    }
+
+    if (Hvel > 0.0)
+    {
+        double VelNorm = std::sqrt(rVelocity[0]*rVelocity[0] + rVelocity[1]*rVelocity[1] + rVelocity[2]*rVelocity[2]);
+        Hvel /= VelNorm;
+    }
+
+    return Hvel;
+}
+
+// Hexahedra3D8 version.
+template<>
+double ElementSizeCalculator<3,8>::ProjectedElementSize(const Geometry<Node<3> >& rGeometry,
+                                                        const array_1d<double,3>& rVelocity) {
+
+    double Hvel = 0.0;
+
+    // Logic: define a box given by hexahedra edges 10,30,40 (which I'm assuming to be orthogonal)
+    // Transform the given direction to the coordinate system defined by the edges.
+    // In this reference frame, place a vector aligned with rVelocity on the origin and see
+    // which side of the box intersects the vector (that is: which component of the vector is larger)
+    // Then scale the vector to hit the limit of the box.
+    // The lenght of the vector is what we are looking for.
+    // NOTE: we use absolute values on all checks, this allows us to simplify the problem
+    // to a single sector (otherwise we would start by determining in which of the eight sectors
+    // given by +-x, +-y, +-z we have to look for the intersection).
+    array_1d<double,3> U = rVelocity;
+    //Normalize U
+    U /= std::sqrt(U[0]*U[0] + U[1]*U[1] + U[2]*U[2]);
+
+    array_1d<double,3> v10 = rGeometry[1].Coordinates() - rGeometry[0].Coordinates();
+    array_1d<double,3> v30 = rGeometry[3].Coordinates() - rGeometry[0].Coordinates();
+    array_1d<double,3> v40 = rGeometry[4].Coordinates() - rGeometry[0].Coordinates();
+
+    // Express U in the coordinate system defined by {v10,v30,v40}
+    Matrix Q = ZeroMatrix(3,3);
+    for (unsigned int i = 0; i < 3; i++)
+    {
+        Q(i,0) = v10[i];
+        Q(i,1) = v30[i];
+        Q(i,2) = v40[i];
+    }
+
+    Matrix QInv = ZeroMatrix(3,3);
+
+    // Invert Matrix Q
+    typedef permutation_matrix<std::size_t> pmatrix;
+    // create a working copy of the input matrix
+    Matrix A(Q);
+    // create a permutation matrix for the LU-factorization
+    pmatrix pm(A.size1());
+    // perform LU-factorization
+    int res = lu_factorize(A,pm);
+    if( res != 0 )
+        return false;
+    // create identity matrix of "inverse"
+    QInv.assign(identity_matrix<double>(A.size1()));
+    // backsubstitute to get the inverse
+    lu_substitute(A, pm, QInv);
```

Comment:
There is a function in math utils that inverts dense matrices. I'd recommend you to use that one, since the code here will probably become obsolete once we remove ublas.